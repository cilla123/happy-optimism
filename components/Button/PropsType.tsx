export default interface PropsType {
  theme?: 'default' | 'primary' | 'success' | 'warning' | 'error';
  size?: 'xl' | 'lg' | 'sm' | 'xs';
  shape?: 'radius' | 'round' | 'circle';
  block?: boolean;
  bordered?: boolean;
  active?: boolean;
  focus?: boolean;
  disabled?: boolean;
  loading?: boolean;
  icon?: JSX.Element;
  onClick?: (e?: any) => void;
}
