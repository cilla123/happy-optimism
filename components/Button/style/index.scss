@import "./_button/_optimism-button-variable.scss";
@import "./_button/_optimism-button.scss";

/* 大小相关 */
@import './_button/_optimism-button-size.scss';

/* 风格相关 */
@import './_button/_optimism-button-style.scss';

/* 状态相关 */
@import './_button/_optimism-button-status.scss';
@import './_button/_optimism-button-bordered.scss';

/* Icon */
@import './_button/_optimism-button-icon.scss';

/* Button Group */
@import './_button/_optimism-button-group.scss';
