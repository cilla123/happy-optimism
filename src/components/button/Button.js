import React, { Component, PureComponent } from 'react'
import setClass from 'classnames'
import Proptypes from 'prop-types'
import omit from 'lodash/omit'
// import Icon from 'icon'

const BLACK_LIST = [
    'type'
]

const BASE_LIST = ['href', 'target']

const BTN_BLACK_LIST = BASE_LIST.concat(BLACK_LIST)

const A_BLACK_LIST = BASE_LIST.concat(BLACK_LIST)

const wrapTextWithSpanTag = children => {
    return React.Children.map(children, child => {
        if (typeof child === 'string') {
            return <span>{child}</span>
        }
        return child
    })
}

export default class Button extends (PureComponent || Component) {
    
    static propTypes = {
        type: Proptypes.oneOf(['default', 'primary', 'success', 'danger', 'link']),
    }

    static defaultProps = {
        type: 'default'
    }

    constructor(props){
        super(props)
    }

    renderLink(){
        return (
            <a>11</a>
        )
    }

    renderButton(classNames, iconNode, wrapedChildren){
        const { component } = this.props
        const Node = component || 'button'
        const nodeProps = omit(this.props, BTN_BLACK_LIST)

        return (
            <Node 
                {...nodeProps}
            >
                { iconNode }
                { wrapedChildren }
            </Node>
        )
    }

    render(){
        const { type, prefix } = this.props
        const { className } = this.props
        const renderer = href || target ? 'renderLink' : 'renderButton'
        const classNames = setClass(
            {
                [`${prefix}-btn-${type}${outline ? '-outline' : ''}`] : type !== 'default',
            },
            `${prefix}-btn`,
            className
        )
        const iconNode = null;
        const wrapedChildren = wrapTextWithSpanTag(children)

        return this[renderer](classNames, iconNode, wrapedChildren)
    }

}
