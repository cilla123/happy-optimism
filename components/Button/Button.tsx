import React, { Component, PureComponent } from 'react'
import PropsType from './PropsType'
import PropTypes from 'prop-types'
import omit from 'lodash/omit'
import setClass from 'classnames'
import Icon from '../Icon'

const BLACK_LIST = [
    'type',
    'size',
    'htmlType',
    'block',
    'component',
    'disabled',
    'loading',
    'outline',
    'bordered',
    'icon',
    'className',
    'prefix',
]

const BTN_BLACK_LIST = ['href', 'target'].concat(BLACK_LIST)

const A_BLACK_LIST = ['href', 'target'].concat(BLACK_LIST)

const wrapTextWithSpanTag = children => {
    return React.Children.map(children, child => {
        return typeof child === 'string' ? <span>{ child }</span> : child
    })
}

class Button extends (PureComponent || Component)<PropsType, any> {

    static propTypes = {
        type: PropTypes.oneOf(['default', 'primary', 'success', 'danger', 'link']),
        size: PropTypes.oneOf(['large', 'medium', 'small']),
        htmlType: PropTypes.oneOf(['button', 'submit', 'reset']),
        className: PropTypes.string,
        block: PropTypes.bool,
        component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
        disabled: PropTypes.bool,
        loading: PropTypes.bool,
        outline: PropTypes.bool,
        bordered: PropTypes.bool,
        prefix: PropTypes.string,
        icon: PropTypes.oneOfType([PropTypes.string, PropTypes.func])
    }
    
    static defaultProps = {
        type: 'default',
        size: 'medium',
        htmlType: 'button',
        className: '',
        block: false,
        disabled: false,
        loading: false,
        outline: false,
        bordered: true,
        prefix: 'happy-optimism',
    }

    constructor(props){
        super(props)
        this.handleClick = this.handleClick.bind(this)
    }

    /**
     * 处理点击事件
     */
    handleClick(event){
        if (this.props.disabled || this.props.loading) return

        if (this.props.onClick) {
            this.props.onClick(event)
        }
    }

    /**
     * 渲染 a 标签
     */
    renderLink(classNames, iconNode, wrapedChildren){
        const { component, disabled, loading, href = '', target } = this.props
        const Node = component ||  'a'
        const nodePorps = omit(this.props, A_BLACK_LIST)

        return (
            <Node
                {...(disabled || loading ? {} : {href, target})}
                {...nodePorps}
                className={classNames}
                onClick={ this.handleClick }
            >
                { iconNode }
                { wrapedChildren }
            </Node>
        )
    }

    /**
     * 渲染 button 标签
     */
    renderButton(classNames, iconNode, wrapedChildren) {
        const { component, disabled, loading, htmlType } = this.props;
        const Node = component || 'button';
        const nodeProps = omit(this.props, BTN_BLACK_LIST);

        return (
            <Node
                {...nodeProps}
                {...(htmlType ? { type: htmlType } : {})}
                className={classNames}
                disabled={disabled || loading}
                onClick={this.handleClick}
            >
                {iconNode}
                {wrapedChildren}
            </Node>
        )
    }

    render() {
        const {
            href,
            target,
            type,
            size,
            block,
            disabled,
            loading,
            outline,
            bordered,
            prefix,
            icon,
            children,
        } = this.props
        let renderer = href || target ? 'renderLink' : 'renderButton'
        let { className } = this.props
        let classNames = setClass(
            {
            [`${prefix}-btn-${type}${outline ? '-outline' : ''}`]:
                type !== 'default',
            [`${prefix}-btn-${size}`]: size !== 'medium',
            [`${prefix}-btn-block`]: block,
            [`${prefix}-btn-loading`]: loading,
            [`${prefix}-btn-disabled`]: disabled,
            [`${prefix}-btn-border-transparent`]: !bordered,
            },
            `${prefix}-btn`,
            className
        );
        
        const iconNode = icon ? <Icon type={icon} /> : null;        
        const wrapedChildren = wrapTextWithSpanTag(children);

        return this[renderer](classNames, iconNode, wrapedChildren);
    }

}

export default Button