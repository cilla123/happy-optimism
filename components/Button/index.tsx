import React, { PureComponent } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';
import Spinner from '../Spinner';

export interface ButtonProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class Button extends PureComponent<ButtonProps, {}> {

  static defaultProps = {
    prefixCls: 'za-button',
    theme: 'default',
    block: false,
    bordered: false,
    active: false,
    focus: false,
    disabled: false,
    loading: false,
    onClick() {},
  };

  render() {
    const {
      prefixCls,
      className,
      theme,
      size,
      shape,
      icon,
      block,
      active,
      focus,
      bordered,
      disabled,
      loading,
      onClick,
      children,
      ...others,
    } = this.props;

    const classes = classnames(`${prefixCls}`, className, {
      [`theme-${theme}`]: !!theme,
      [`size-${size}`]: !!size,
      [`shape-${shape}`]: !!shape,
      block,
      bordered,
      active,
      focus,
      disabled,
    });

    const iconRender = loading
      ? <Spinner className="rotate360" />
      : icon;

    const childrenRender = children && <span>{children}</span>;

    const contentRender = (!!icon || loading)
      ? <div className={`${prefixCls}-content`}>{iconRender}{childrenRender}</div>
      : childrenRender;

    return (
      <a
        role="button"
        aria-disabled={disabled}
        className={classes}
        onClick={e => !disabled && typeof onClick === 'function' && onClick(e)}
        onTouchStart={() => {}}
        {...others}
      >
        {contentRender}
      </a>
    );
  }
}
