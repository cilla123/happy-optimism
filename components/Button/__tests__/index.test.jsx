import React from 'react';
import { render, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Button from '../index';

describe('Button', () => {
  it('renders correctly', () => {
    const wrapper = render(<Button>foo</Button>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('theme', () => {
    const wrapper = render(<Button theme="primary">foo</Button>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('block', () => {
    const wrapper = render(<Button block>foo</Button>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('bordered', () => {
    const wrapper = render(<Button bordered>foo</Button>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('size', () => {
    const wrapper = render(<Button size="lg">foo</Button>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('shape is radius', () => {
    const wrapper = render(<Button shape="radius">foo</Button>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('icon', () => {
    const wrapper = render(<Button icon={<img alt="" src="https://zhongantecheng.github.io/zarm/images/state.18e78939.png" />}>foo</Button>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('loading', () => {
    const wrapper = render(<Button loading>foo</Button>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('disabled', () => {
    const wrapper = render(<Button disabled>foo</Button>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('onClick', () => {
    const onClick = jest.fn();
    const wrapper = shallow(<Button onClick={onClick}>foo</Button>);
    wrapper.simulate('click');
    expect(onClick).toBeCalled();
  });
});
