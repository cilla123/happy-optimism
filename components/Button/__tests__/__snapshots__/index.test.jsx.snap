// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button block 1`] = `
<a
  aria-disabled="false"
  class="za-button theme-default block"
  role="button"
>
  <span>
    foo
  </span>
</a>
`;

exports[`Button bordered 1`] = `
<a
  aria-disabled="false"
  class="za-button theme-default bordered"
  role="button"
>
  <span>
    foo
  </span>
</a>
`;

exports[`Button disabled 1`] = `
<a
  aria-disabled="true"
  class="za-button theme-default disabled"
  role="button"
>
  <span>
    foo
  </span>
</a>
`;

exports[`Button icon 1`] = `
<a
  aria-disabled="false"
  class="za-button theme-default"
  role="button"
>
  <div
    class="za-button-content"
  >
    <img
      alt=""
      src="https://zhongantecheng.github.io/zarm/images/state.18e78939.png"
    />
    <span>
      foo
    </span>
  </div>
</a>
`;

exports[`Button loading 1`] = `
<a
  aria-disabled="false"
  class="za-button theme-default"
  role="button"
>
  <div
    class="za-button-content"
  >
    <svg
      class="za-spinner rotate360 theme-primary"
      viewBox="0 0 62 62"
    >
      <circle
        class="za-spinner-path"
        cx="31"
        cy="31"
        fill="none"
        r="28.5"
        style="stroke-width:5"
      />
      <circle
        class="za-spinner-line"
        cx="31"
        cy="31"
        fill="none"
        r="28.5"
        style="stroke-dasharray:26.860617188192727 179.0707812546182;stroke-width:5"
      />
    </svg>
    <span>
      foo
    </span>
  </div>
</a>
`;

exports[`Button renders correctly 1`] = `
<a
  aria-disabled="false"
  class="za-button theme-default"
  role="button"
>
  <span>
    foo
  </span>
</a>
`;

exports[`Button shape is radius 1`] = `
<a
  aria-disabled="false"
  class="za-button theme-default shape-radius"
  role="button"
>
  <span>
    foo
  </span>
</a>
`;

exports[`Button size 1`] = `
<a
  aria-disabled="false"
  class="za-button theme-default size-lg"
  role="button"
>
  <span>
    foo
  </span>
</a>
`;

exports[`Button theme 1`] = `
<a
  aria-disabled="false"
  class="za-button theme-primary"
  role="button"
>
  <span>
    foo
  </span>
</a>
`;
